### PR TITLE
Cache external git call results

### PR DESCRIPTION
**Pull request by flaviut**
### What does this PR do?

It's expensive to call programs. And with the previous implementation,
git would be called a few times for each test, which causes dramatic
slowdowns when there are thousands of tests.

We can safely memoize these values. It would be incredibly unusual to
have the current version of the code changed out from under us, and in
that circumstance, we already cannot provide any sort of reasonable
behavior.

### Motivation

Adding the dd-trace agent to our tests in CI caused massive slowdowns:

2022-07-19-145724_1720x1279_scrot

With this PR in place, there's no longer any obvious hot spot:

2022-07-19-150505_1836x1228_scrot

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes

Not sure if we get any special treatment, but I'm submitting this on behalf of ONE.

### CODE REVIEW

**Comment by rochdev** [comment]
> And with the previous implementation, git would be called a few times for each test, which causes dramatic
slowdowns when there are thousands of tests

@juan-fernandez Doesn't this actually happen only once per plugin?

Generally speaking I agree with the idea of memoizing the result, although we should also make sure that it doesn't happen more often than it should regardless of any optimizations. It would also make sense I think to make sure that the function runs only when the plugin is **enabled** and not just when it's **initialized**, to avoid calling it for CI integrations that are not even used. This is especially true since it's highly unlikely that multiple test runners are used at the same time.

File: packages/dd-trace/src/plugins/util/git.js [hunk 1]
```diff
@@ -15,8 +15,15 @@ const {
   CI_WORKSPACE_PATH
 } = require('./tags')
 
+const cache = new Map()
+
 // If there is ciMetadata, it takes precedence.
 function getGitMetadata (ciMetadata) {
+  const cacheKey = JSON.stringify(ciMetadata)
+  if (cache.has(cacheKey)) {
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/git.js:23]: It would be more efficient to `get` right away and check if there is a value to avoid 2 lookups.

**Comment by rochdev** [comment]
@juan-fernandez Do we have any existing benchmarks that would help quantify the difference and avoid future regressions?

**Comment by juan-fernandez** [comment]
> > And with the previous implementation, git would be called a few times for each test, which causes dramatic
> > slowdowns when there are thousands of tests
> 
> @juan-fernandez Doesn't this actually happen only once per plugin?
> 
> Generally speaking I agree with the idea of memoizing the result, although we should also make sure that it doesn't happen more often than it should regardless of any optimizations. It would also make sense I think to make sure that the function runs only when the plugin is **enabled** and not just when it's **initialized**, to avoid calling it for CI integrations that are not even used. This is especially true since it's highly unlikely that multiple test runners are used at the same time.

yeah it _should_ happen only once per plugin. I'll check that. 

Also, agreed with the point: we should only run git commands if the plugin is being used, not just initialized. I'll look into that.

**Comment by juan-fernandez** [comment]
> I figured it out--I was using it wrong.
> 
> We have a .js file that runs `dd.init`, since we want to customize things a bit. I put this file in [jest's `setupFiles` field](https://jestjs.io/docs/configuration#setupfiles-array). I thought that meant that it only gets run once, but those files actually get executed once per test.
> 
> Functional alternatives:
> 
> * [using `node -r`](https://nodejs.org/api/cli.html#-r---require-module)
> * [using jest's `globalSetup`](https://jestjs.io/docs/configuration#globalsetup-string)
> 
> Leaving this open in case you want to use it to track something else, but things are all resolved on my side.

Thanks for the update! I think this change still makes sense regardless, so we're working on defining how to land this or something similar to this.

**Comment by rochdev** [comment]
This is not the first time that `init` is called in a hot path from misuse, so we should probably add a benchmark looping through the method to ensure that it doesn't have too big of a performance impact to avoid causing issues. Basically, misuse of dd-trace should only cause issues with dd-trace, not with the application.

**Comment by tlhunter** [comment]
@juan-fernandez are you tracking this change request somewhere else? And if so is it okay if I close this PR?

**Comment by juan-fernandez** [comment]
let's close it

**Pull request closed without merging**